specs/Spec.AES.fst

```diff
@@ -423,3 +436,179 @@ let aes128_ctr_key_block1 key n_len n =
+  let key_ex = aes_key_expansion_LE v k in
+  { key_ex = key_ex; block = input}
+
+let aes_ctr32_encrypt_block_LE
+  (v:variant)
+  (key:aes_xkey v)
+  (n:block)
+  (incr:uint32)
+  (b:block) :
+  Tot block =
+
+  let n' = aes_ctr32_set_counter_LE n incr in
+  let kb = aes_encrypt_block v key n' in
+  reverse_bytes (map2 (^.) (reverse_bytes b) kb)
+
+let aes_ctr32_encrypt_block_4_LE
+  (v:variant)
+  (key:aes_xkey v)
+  (n:block)
+  (incr:uint32)
+  (b:lbytes 64):
+  Tot (lbytes 64) =
+
+  let o0 = aes_ctr32_encrypt_block_LE v key n
+    incr (sub b 0 16) in
+  let o1 = aes_ctr32_encrypt_block_LE v key n
+    (incr +. u32 1) (sub b 16 16) in
+  let o2 = aes_ctr32_encrypt_block_LE v key n
+    (incr +. u32 2) (sub b 32 16) in
+  let o3 = aes_ctr32_encrypt_block_LE v key n
+    (incr +. u32 3) (sub b 48 16) in
+  concat (concat (concat o0 o1) o2) o3
+
+let aes_ctr32_encrypt_last_LE
+  (v:variant)
+  (key:aes_xkey v)
+  (n:block)
+  (incr:uint32)
+  (len:size_nat{len < 16})
+  (b:lbytes len):
+  Tot (lbytes len) =
+
+  let plain = create 16 (u8 0) in
+  let plain = update_sub plain 0 (length b) b in
+  let cipher = aes_ctr32_encrypt_block_LE v key n incr plain in
+  sub cipher 0 (length b)
+
+let aes_ctr32_encrypt_last_4_LE
+  (v:variant)
+  (key:aes_xkey v)
+  (n:block)
+  (incr:uint32)
+  (len:size_nat{len < 64})
+  (b:lbytes len):
+  Tot (lbytes len) =
+
+  let last = create 64 (u8 0) in
+  let last = update_sub last 0 (length b) b in
+  let last = aes_ctr32_encrypt_block_4_LE v key n incr last in
+  sub last 0 (length b)
+
+let aes_ctr32_encrypt_block_incr_4_LE
+  (v:variant)
+  (key:aes_xkey v)
+  (n:block)
+  (counter:uint32)
+  (incr:size_nat)
+  (b:lbytes 64):
+  Tot (lbytes 64) =
+
+  aes_ctr32_encrypt_block_4_LE v key n (counter +. (u32 incr *. u32 4)) b
+
+let aes_ctr32_encrypt_last_incr_4_LE
+  (v:variant)
+  (key:aes_xkey v)
+  (n:block)
+  (counter:uint32)
+  (incr:size_nat)
+  (len:size_nat{len < 64})
+  (b:lbytes len):
+  Tot (lbytes len) =
+
+  aes_ctr32_encrypt_last_4_LE v key n (counter +. (u32 incr *. u32 4)) len b
+
+val aes_ctr32_encrypt_stream_LE:
+    v:variant
+  -> key:aes_xkey v
+  -> n:block
+  -> counter:uint32
+  -> msg:bytes{length msg / 64 <= max_size_t} ->
+  Tot (ciphertext:bytes{length ciphertext == length msg})
+let aes_ctr32_encrypt_stream_LE v key n counter msg =
+  map_blocks 64 msg
+    (aes_ctr32_encrypt_block_incr_4_LE v key n counter)
+    (aes_ctr32_encrypt_last_incr_4_LE v key n counter)
+
+val aes_ctr32_encrypt_bytes_LE:
+    v:variant
+  -> key:aes_key v
+  -> nonce:lbytes 12
+  -> c:uint32
+  -> msg:bytes{length msg / 64 <= max_size_t} ->
+  Tot (ciphertext:bytes{length ciphertext == length msg})
+
+let aes_ctr32_encrypt_bytes_LE v key nonce ctr0 msg =
+  let st0 = aes_ctr32_init_LE v key nonce in
+  aes_ctr32_encrypt_stream_LE v st0.key_ex
+    st0.block ctr0 msg
+
+let aes_ctr32_decrypt_bytes_LE v key nonce ctr0 msg =
+  aes_ctr32_encrypt_bytes_LE v key nonce ctr0 msg
```

Comment:
where is the proof that encrypt (decrypt x) == x?

Reply:
that proof should be applied for `aes_ctr32_decrypt_bytes_LE` function?

Reply:
yes -- it should be pretty trivial in this case but it's a good "sanity check" to make sure you can prove that?

Reply:
Now I'm thinking about it, there shouldn't have been `aes_ctr32_decrypt` function. GCM uses the forward version of AES cipher for encryption and decryption. I checked Vale AES/GCTR spec and there's no decryption function. See https://github.com/hacl-star/hacl-star/blob/main/vale/specs/crypto/Vale.AES.AES_s.fst
https://github.com/hacl-star/hacl-star/blob/main/vale/specs/crypto/Vale.AES.GCTR_s.fst

Reply:
Vale has no such lemma, in fact Vale doesn't have decryption function for AES/GCTR spec. See  https://github.com/hacl-star/hacl-star/blob/main/vale/specs/crypto/Vale.AES.AES_s.fst
https://github.com/hacl-star/hacl-star/blob/main/vale/specs/crypto/Vale.AES.GCTR_s.fst
GCM uses the forward version of AES cipher for encryption and decryption, so I think it's good to have `aes_ctr32_decrypt` unlike Vale but a lemma for encrypt (decrypt x) == x is to verify the spec itself and not the implementation.